clients/src/main/java/org/apache/kafka/common/internals/FetchBuilder.java

```diff
@@ -0,0 +1,170 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * This builder is a useful building block for doing fetch requests where topic partitions have to be rotated via
+ * round-robin to ensure fairness and some level of determinism given the existence of a limit on the fetch response
+ * size. Because the serialization of fetch requests is more efficient if all partitions for the same topic are grouped
+ * together, we do such grouping in the method `set`.
+ *
+ * As partitions are moved to the end, the same topic may be repeated more than once.
+ */
+public class FetchBuilder<S> {
+
+    private final LinkedHashMap<TopicPartition, S> map = new LinkedHashMap<>();
+
+    public FetchBuilder() {}
+
+    public void moveToEnd(TopicPartition topicPartition) {
+        S state = map.remove(topicPartition);
+        if (state != null)
+            map.put(topicPartition, state);
+    }
+
+    public void updateAndMoveToEnd(TopicPartition topicPartition, S state) {
+        map.remove(topicPartition);
+        map.put(topicPartition, state);
+    }
+
+    public void remove(TopicPartition topicPartition) {
+        map.remove(topicPartition);
+    }
+
+    /**
+     * Returns the partitions in random order.
+     */
+    public Set<TopicPartition> partitionSet() {
+        return map.keySet();
+    }
+
+    public void clear() {
+        map.clear();
+    }
+
+    public boolean contains(TopicPartition topicPartition) {
+        return map.containsKey(topicPartition);
+    }
+
+    /**
+     * Returns the partition states in order.
+     */
+    public List<PartitionState<S>> partitionStates() {
+        List<PartitionState<S>> result = new ArrayList<>();
+        for (Map.Entry<TopicPartition, S> entry : map.entrySet()) {
+            result.add(new PartitionState<S>(entry.getKey(), entry.getValue()));
+        }
+        return result;
+    }
+
+    /**
+     * Returns the partition state values in order.
+     */
+    public List<S> partitionStateValues() {
+        return new ArrayList<>(map.values());
+    }
+
+    public S stateValue(TopicPartition topicPartition) {
+        return map.get(topicPartition);
+    }
+
+    public int size() {
+        return map.size();
+    }
+
+    /**
+     * Update the builder to have the received map as its state (i.e. the previous state is cleared). The builder will
+     * "batch by topic", so if we have a, b and c, each with two partitions, we may end up with something like the
+     * following (the order of topics and partitions within topics is dependent on the iteration order of the received
+     * map): a0, a1, b1, b0, c0, c1.
+     */
+    public void set(Map<TopicPartition, S> partitionToState) {
+        map.clear();
+        update(partitionToState);
+    }
+
+    private void update(Map<TopicPartition, S> partitionToState) {
+        LinkedHashMap<String, List<TopicPartition>> topicToPartitions = new LinkedHashMap<>();
```

Comment:
I specified that we respect the order of `partitionToState`. Makes it a bit easier to write tests and seemed a little nicer.